### PR TITLE
:bug: Add SecurityGroup to Loadbalancerport when not using octavia

### LINKED ISSUE #694: missing SecurityGroup on neutron API-Loadbalancer
/kind bug

Hi.

**What steps did you take and what happened:**
I deployed an an k8s-cluster via the openstack-api-provider and external-cloud-provider with the following configuration:

```
  managedAPIServerLoadBalancer: true
  managedSecurityGroups: true
  nodeCidr: 10.6.0.0/24
  useOctavia: false
```

The loadblanacer and corresponding pools, members, monitors and listeners were created and the first control-plane-node starts  and bootstraps successfully. But my kind based bootstrap-cluster can't reach the api of the newly created node. After updating the default-security-group and adding port 6443/tcp the capo-provider was able to proceed and fetch its information about the cluster-status.

**What did you expect to happen:**
I expect to get access to api via loadbalancer directly, without doing manual steps and adding SecurityGroupRules to the default-group.

**Anything else you would like to add:**
For me it seems to be ok if the control-plane-securitygroup will be assigned to the loadbalancer-port, or a dedicated securitygroup which allows access to the api-port and maybe apiServerLoadBalancerAdditionalPorts. Additionally the default-group can be removed from the loadbalancer-port.

Or is the usage of neutron-lbaas obsolet and i should use octavia? Is there a workaround for that problem?

**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built): 
```
cluster-api-bootstrap:~# kubectl get pods -n capo-system -o jsonpath='{ .items[*].spec.containers[*].image }'
gcr.io/kubebuilder/kube-rbac-proxy:v0.4.1 us.gcr.io/k8s-artifacts-prod/capi-openstack/capi-openstack-controller:v0.3.3
```

- Cluster-API version:
```
cluster-api-bootstrap:~# clusterctl version
clusterctl version: &version.Info{Major:"0", Minor:"3", GitVersion:"v0.3.11", GitCommit:"e9cf6846b6d93dedadfcf44c00357d15f5ccba64", GitTreeState:"clean", BuildDate:"2020-11-19T18:49:17Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
cluster-api-bootstrap:~# kubectl get pods -n capi-system -o jsonpath='{ .items[*].spec.containers[*].image }'
gcr.io/kubebuilder/kube-rbac-proxy:v0.4.1 us.gcr.io/k8s-artifacts-prod/cluster-api/cluster-api-controller:v0.3.11
```

- Minikube/KIND version:
```
cluster-api-bootstrap:~# kind --version
kind version 0.9.0
```

- Kubernetes version (use `kubectl version`):
```
cluster-api-bootstrap:~# kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:30:33Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"clean", BuildDate:"2020-09-14T07:30:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```

- OS (e.g. from `/etc/os-release`):
```
cluster-api-bootstrap:~# cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

Thanks for the great work.

Regards
Tino

### PR DESCRIPTION
**Pull request by zoltan0907**
**What this PR does / why we need it**:
Make API usable when using Neutron Lbaas and managedAPIServerLoadbalancer

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #694

**Special notes for your reviewer**:

**Release note**:
Adds controlplaneSecurityGroup to LoadbalancerPort when using Neutron Lbaas.

### CODE REVIEW

**Comment by zoltan0907** [comment]
I signed it

**Comment by zoltan0907** [comment]
I signed it

**Comment by zoltan0907** [comment]
There is an error in the openlab/check, but i'm unsure if that error is related to my changes. The controlplane was provisioned and cni was deployed. Additionally i looked int othe [capo-logs](https://logs.openlabtesting.org/logs/00/700/18e2826409fe765057f2f502f5eef7269cb85482/check/cluster-api-provider-openstack-current-acceptance-test-v1.18/254792a/logs/_artifacts/logs/clusterapi-control-plane/pods/capo-system_capo-controller-manager-7bd8478cb6-lbcl6_4bcbc9de-7aeb-4b25-884e-1152201b87b1/manager/0.log) and saw that `managedAPIServerLoadBalancer=false`.

**Comment by jichenjc** [comment]
@zoltan0907 just FYI, the check is not 100% success due to our CI env is not stable, @sbueringer submitted a PR
with the openlab (CI env) and so far still not fixed

so please leave a 'recheck' if you think it's not the code's failure and usually 1/2 rate the test pass

**Comment by jichenjc** [comment]
looks reasonable
need fix ut now ..

**Comment by hidekazuna** [comment]
@zoltan0907 

> Adds controlplaneSecurityGroup to LoadbalancerPort when using Neutron Lbaas.

The security group for vip port opens [openStackCluster.Spec.APIServerPort](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/api/v1alpha3/openstackcluster_types.go#L82) and [openStackCluster.Spec.APIServerLoadBalancerAdditionalPorts](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/api/v1alpha3/openstackcluster_types.go#L85) only. We will create new security group for vip port and add it.

File: pkg/cloud/services/loadbalancer/loadbalancer.go [hunk 1]
```diff
@@ -61,6 +65,35 @@ func (s *Service) ReconcileLoadBalancer(clusterName string, openStackCluster *in
 		return err
 	}
 
+	if !s.usesOctavia {
```
Comment (hidekazuna) [inline pkg/cloud/services/loadbalancer/loadbalancer.go:68]: We can use `openStackCluster.Spec.UseOctavia`.

File: pkg/cloud/services/networking/securitygroups.go [hunk 2]
```diff
@@ -26,10 +26,11 @@ import (
...
 const (
-	secGroupPrefix     string = "k8s"
-	controlPlaneSuffix string = "controlplane"
-	workerSuffix       string = "worker"
-	bastionSuffix      string = "bastion"
+	SecGroupPrefix     string = "k8s"
+	ControlPlaneSuffix string = "controlplane"
+	WorkerSuffix       string = "worker"
+	BastionSuffix      string = "bastion"
+	NeutronLbaasSuffix string = "lbaas"
```
Comment (hidekazuna) [inline pkg/cloud/services/networking/securitygroups.go:33]: An unexported constant is `secGroupPrefix` not `SecGroupPrefix`

https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps

File: pkg/cloud/services/loadbalancer/loadbalancer.go [hunk 3]
```diff
@@ -61,6 +65,35 @@ func (s *Service) ReconcileLoadBalancer(clusterName string, openStackCluster *in
...
+
+		if len(neutronLbaasGroups) != 1 {
+			return fmt.Errorf("error found %v securitygroups with name %v", len(neutronLbaasGroups), neutronLbaasSecGroupName)
+		}
+
+		updateOpts := v2_ports.UpdateOpts{
+			SecurityGroups: &[]string{neutronLbaasGroups[0].ID},
+		}
+
+		_, err = v2_ports.Update(s.networkingClient, lb.VipPortID, updateOpts).Extract()
```
Comment (hidekazuna) [inline pkg/cloud/services/loadbalancer/loadbalancer.go:91]: Could you make function?

**Review by hidekazuna — CHANGES_REQUESTED** [review]

File: pkg/cloud/services/loadbalancer/loadbalancer.go [hunk 4]
```diff
@@ -19,17 +19,21 @@ package loadbalancer
...
+
 	"github.com/gophercloud/gophercloud"
 	"github.com/gophercloud/gophercloud/openstack/loadbalancer/v2/listeners"
 	"github.com/gophercloud/gophercloud/openstack/loadbalancer/v2/loadbalancers"
 	"github.com/gophercloud/gophercloud/openstack/loadbalancer/v2/monitors"
 	"github.com/gophercloud/gophercloud/openstack/loadbalancer/v2/pools"
 	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/layer3/floatingips"
-	"k8s.io/apimachinery/pkg/util/wait"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/security/groups"
+	v2_ports "github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
```
Comment (hidekazuna) [inline pkg/cloud/services/loadbalancer/loadbalancer.go:34]: v2_ports is not necessary.

File: pkg/cloud/services/networking/securitygroups.go (hunk 2 above)
Comment (zoltan0907) [reply pkg/cloud/services/networking/securitygroups.go:33]: SecGroupPrefix is used in https://github.com/zoltan0907/cluster-api-provider-openstack/blob/master/pkg/cloud/services/loadbalancer/loadbalancer.go#L181

File: pkg/cloud/services/loadbalancer/loadbalancer.go (hunk 3 above)
Comment (hidekazuna) [reply pkg/cloud/services/loadbalancer/loadbalancer.go]: Thanks. confirmed.

File: pkg/cloud/services/networking/securitygroups.go (hunk 2 above)
Comment (hidekazuna) [reply pkg/cloud/services/networking/securitygroups.go:33]: This is about naming convention.
Check out https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps

File: pkg/cloud/services/loadbalancer/loadbalancer.go (hunk 1 above)
Comment (hidekazuna) [reply pkg/cloud/services/loadbalancer/loadbalancer.go]: Thanks, confirmed.

File: pkg/cloud/services/loadbalancer/loadbalancer.go (hunk 4 above)
Comment (hidekazuna) [reply pkg/cloud/services/loadbalancer/loadbalancer.go]: Thanks, confirmed.

File: pkg/cloud/services/networking/securitygroups.go (hunk 2 above)
Comment (zoltan0907) [reply pkg/cloud/services/networking/securitygroups.go]: Sadly i don't understand what you want me todo. Should `SecGroupPrefix` transformed into `SECGroupPrefix` or shoud i add a comment to the export? I rewrote this peace of code and made a exported function to calculate the name of the securitygroup. Hopefully this is sufficient.

File: pkg/cloud/services/networking/securitygroups.go [hunk 5]
```diff
@@ -25,12 +25,15 @@ import (
...
-	secGroupPrefix     string = "k8s"
-	controlPlaneSuffix string = "controlplane"
-	workerSuffix       string = "worker"
-	bastionSuffix      string = "bastion"
-	remoteGroupIDSelf  string = "self"
+	secGroupPrefix                string = "k8s"
+	ControlPlaneSuffix            string = "controlplane"
+	WorkerSuffix                  string = "worker"
+	BastionSuffix                 string = "bastion"
+	NeutronLbaasSuffix            string = "lbaas"
```
Comment (hidekazuna) [inline pkg/cloud/services/networking/securitygroups.go:34]: ControlPlaneSuffix, WorkerSuffix, BastionSuffix, NeutronLbaasSuffix should start with small character: controlPlaneSuffix, workerSuffix, bastionSuffix, neutronLbaasSuffix respectively.

File: pkg/cloud/services/networking/securitygroups.go [hunk 6]
```diff
@@ -25,12 +25,15 @@ import (
 	"sigs.k8s.io/cluster-api-provider-openstack/pkg/record"
 )
 
+// Exported Constants can be used to calculate securitygroupnames in other packages
```
Comment (hidekazuna) [inline pkg/cloud/services/networking/securitygroups.go:28]: This PR focus on fixing the bug, not update variable name. If you want to refactor, raise another PR, please.

**Review by hidekazuna — CHANGES_REQUESTED** [review]

File: pkg/cloud/services/networking/securitygroups.go (hunk 5 above)
Comment (hidekazuna) [reply pkg/cloud/services/networking/securitygroups.go:34]: NeutronLbaasSuffix is exported to loadbalancer package. ControlPlaneSuffix, WorkerSuffix, BastionSuffix are not exported in my understanding. Let me know if I am wrong.

File: pkg/cloud/services/networking/securitygroups.go (hunk 2 above)
Comment (hidekazuna) [reply pkg/cloud/services/networking/securitygroups.go]: NeutronLbaasSuffix is used in https://github.com/zoltan0907/cluster-api-provider-openstack/blob/master/pkg/cloud/services/loadbalancer/loadbalancer.go#L181. I can not find SecGroupPrefix in the line.

File: pkg/cloud/services/loadbalancer/loadbalancer.go [hunk 7]
```diff
@@ -166,6 +177,36 @@ func (s *Service) ReconcileLoadBalancer(clusterName string, openStackCluster *in
 	return nil
 }
 
+func (s *Service) assignNeutronLbaasAPISecGroup(clusterName string, lb *loadbalancers.LoadBalancer) error {
+	neutronLbaasSecGroupName := fmt.Sprintf("%s-cluster-%s-secgroup-%s", networking.SecGroupPrefix, clusterName, networking.NeutronLbaasSuffix)
```
Comment (hidekazuna) [inline pkg/cloud/services/loadbalancer/loadbalancer.go:181]: We can do not exporting SecGroupPrefix and NeutronLbaasSuffix.


This line:
neutronLbaasSecGroupName := networking.GetNeutronLBaasSecGroupName(clusterName)


Add a following function in pkg/cloud/services/networking/securitygroups.go


```go

func GetNeutronLBaasSecGroupName(clusterName string) string {
 return fmt.Sprintf("%s-cluster-%s-secgroup-%s", secGroupPrefix, clusterName, neutronLbaasSuffix)
}

**Comment by hidekazuna** [comment]
@zoltan0907 Thanks. All fine to me.

**Pull request merged**